### PR TITLE
Server Reflection

### PR DESCRIPTION
**Pull request by mholo65**
Server reflection wip...

Needs https://github.com/protobuf-net/protobuf-net/pull/617

### CODE REVIEW

**Comment by mgravell** [comment]
Is .ionide/symbolCache.db meant to be part of the commit? Or should that be added to .gitignore?

**Comment by mholo65** [comment]
> Is .ionide/symbolCache.db meant to be part of the commit?

No, that's just for the F# extension for VS Code. I'll remove and add to .gitignore.

File: src/protobuf-net.Grpc.Reflection/FileDescriptorSetFactory.cs [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+                    Name = serviceName + ".proto",
+                    Services = { serviceDescriptor },
+                    Syntax = "proto3",
+                    Package = serviceContract.Namespace
+                };
+
+                foreach (var dependency in dependencies)
+                {
+                    fileDescriptor.Dependencies.Add(dependency);
+                    fileDescriptor.AddImport(dependency, true, default);
```
Comment (mholo65) [inline src/protobuf-net.Grpc.Reflection/FileDescriptorSetFactory.cs]: I noticed I needed to `AddImport` in order for `FileDescriptorSet.Process()` to resolve types from dependencies. Shouldn't it be enough to add the dependency to `Dependencies`?

**Comment by mgravell** [comment]
I've merged in the protobuf-net bits; what is still needed on this? I see it is "draft" etc

**Comment by mholo65** [comment]
Thanks! I'll need to update this PR and see if anything still is missing. I'll get back to you once I get to it.

**Comment by mgravell** [comment]
I took long enough, so... yeah, take your time!

**Comment by mgravell** [comment]
Tangential thought: I should update "protogen" to be able to *consume* this API (probably just on netcoreapp3.1) and emit code-gen from an existing service.

**Comment by mholo65** [comment]
Great idea! I also have a PoC for a .NET gRPC client, available as .NET Global Tool, which should be able to consume this, and e.g. generate .proto files. https://github.com/mholo65/dotnet-grpc-cli

**Comment by mgravell** [comment]
Protogen already works from FileDescriptorSet *anyway*, so... Not a huge
leap for me. Hope I'm not treading on your toes if I enable protogen to go
from this API.

**Comment by mholo65** [comment]
No, not at all. My idea of getting Reflection API working with protobuf-net.Grpc is to make interop with other languages easier when using protobuf-net.Grpc with code-first approach.

**Comment by mgravell** [comment]
Yep. Did you see the other thread cross-referenced here? I'm working on the
"generate .proto from services" piece in parallel.

**Comment by mholo65** [comment]
Ah, totally missed that part. Anyway, no issue for me. Haven't put much time on my tool yet, it was just an experiment, since the official `grpc_cli` (https://github.com/grpc/grpc/blob/master/doc/command_line_tool.md) isn't distributed as binary. My idea for the tool, was to be par with `grpc_cli` and additionally add `.proto` generation. Let's see if it's worth it, if `protogen` will also support it. But to answer your question, no, definitely not stepping on any toes :smile:

**Comment by mholo65** [comment]
@mgravell ready for review. I'll look at adding some more tests.

Quickly tested using `Server_CS` sample project and my PoC `grpc_cli`
```
C:\src\gh\dotnet-grpc-cli\grpc.cli [master ≡ +1 ~1 -0 !]> dotnet run dump localhost:5001 Shared_CS.Calculator
syntax = "proto3";
package = Shared_CS;
message MultiplyRequest {
  int32 X = 1;
  int32 Y = 2;
}


syntax = "proto3";
package = Shared_CS;
message MultiplyResult {
  int32 Result = 1;
}


syntax = "proto3";
package = Shared_CS;
service Calculator {
   rpc Multiply(MultiplyRequest) returns (MultiplyResult);
}
```

```
C:\src\gh\dotnet-grpc-cli\grpc.cli [master ≡ +1 ~1 -0 !]> dotnet run dump localhost:5001 MegaCorp.TimeService
syntax = "proto3";
package = ProtoBuf.Grpc.Internal;
message Empty {
}


syntax = "proto3";
package = google.protobuf;
message Timestamp {
  int64 seconds = 1;
  int32 nanos = 2;
}


syntax = "proto3";
package = MegaCorp;
message TimeResult {
  Timestamp Time = 1;
}


syntax = "proto3";
package = MegaCorp;
service TimeService {
   rpc Subscribe(Empty) returns (stream TimeResult);
}
```

**Comment by mgravell** [comment]
This is looking great to me! Merging, many thanks.

**Comment by mgravell** [comment]
FYI I added a few tweaks here; nothing huge, but just tidying up: https://github.com/protobuf-net/protobuf-net.Grpc/commit/4ed34e49fc3ededce6fff495cfa9a40e08056981

**Comment by mholo65** [comment]
Thank you! This was a fun PR!

**Comment by mgravell** [comment]
This will possibly be of interest to you: https://github.com/protobuf-net/protobuf-net/pull/676

What this means is that we can now - whether using a "vanilla" gRPC server, or the code you've just added, do something like:

```
> protogen --grpc https://blah list
Requesting gRPC service directory from 'http://localhost:10042'...
- Shared_CS.Calculator
- MegaCorp.TimeService
- grpc.reflection.v1alpha.ServerReflection
gRPC services discovered: 3

> protogen --grpc get http://localhost:10042 grpc.reflection.v1alpha.ServerReflection --csharp_out=c:\Code\Dump +services=yes
Requesting gRPC 'grpc.reflection.v1alpha.ServerReflection' service from 'http://localhost:10042'...
gRPC descriptors fetched: 3
error 2: type not found: '.grpc.reflection.v1alpha.ServerReflectionRequest'
error 2: type not found: '.grpc.reflection.v1alpha.ServerReflectionResponse'
generated: c:\Code\Dump\grpc.reflection.v1alpha.ServerReflectionRequest.cs
generated: c:\Code\Dump\grpc.reflection.v1alpha.ServerReflectionResponse.cs
generated: c:\Code\Dump\grpc.reflection.v1alpha.ServerReflection.cs
```

(as you can see, it isn't quite perfect yet, but: it is getting close; I think the files are a bit too fragmented currently - possibly related to your comment about imports)

**Comment by mholo65** [comment]
Interesting! Glad to see you already found found a use case for gRPC server reflection!

**Comment by mholo65** [comment]
One thought, in my tool, I'm using [FileDescriptor.BuildFromByteStrings](https://github.com/protocolbuffers/protobuf/blob/master/csharp/src/Google.Protobuf/Reflection/FileDescriptor.cs#L485) from Google.Protobuf, and it expects dependencies to be in correct order. This works fine, so I don't think it's a problem that the FileDescriptors are fragmented. Don't know why you arr seeing the `type not found` errors in `protogen` as FileDescriptors should come in correct order from the reflection service.

**Pull request merged**